src/Symfony/Component/DependencyInjection/Loader/Configurator/ContainerConfigurator.php

```diff
@@ -0,0 +1,175 @@
+ * @method IteratorArgument iterator(array $values)
+ */
+class ContainerConfigurator extends AbstractConfigurator
+{
+    const FACTORY = 'container';
+
+    private $file;
+    private $container;
+    private $loader;
+
+    public function __construct($file, ContainerBuilder $container, PhpFileLoader $loader, &$instanceof)
+    {
+        $this->file = $file;
+        $this->container = $container;
+        $this->loader = $loader;
+        $this->instanceof = &$instanceof;
+    }
+
+    final public function extension($namespace, array $config)
+    {
+        if (!$this->container->hasExtension($namespace)) {
+            $extensions = array_filter(array_map(function ($ext) { return $ext->getAlias(); }, $this->container->getExtensions()));
+            throw new InvalidArgumentException(sprintf(
+                'There is no extension able to load the configuration for "%s" (in %s). Looked for namespace "%s", found %s',
+                $namespace,
+                $this->file,
+                $namespace,
+                $extensions ? sprintf('"%s"', implode('", "', $extensions)) : 'none'
+            ));
+        }
+
+        $this->container->loadFromExtension($namespace, PhpFileLoader::processValue($config));
+    }
+
+    final public function import($resource, $type = null, $ignoreErrors = false)
+    {
+        $this->loader->import($resource, $type, $ignoreErrors, $this->file);
+        $this->loader->setCurrentDir(dirname($this->file));
+    }
+
+    /**
+     * @return ParametersConfigurator
+     */
+    final public function parameters()
+    {
+        return new ParametersConfigurator($this->container);
+    }
+
+    /**
+     * @return ServicesConfigurator
+     */
+    final public function services()
+    {
+        return new ServicesConfigurator($this->container, $this->loader, $this->instanceof);
+    }
+
+    /**
+     * Creates a service reference.
+     *
+     * @param string $id
+     *
+     * @return ReferenceConfigurator
+     */
+    final public function reference($id)
+    {
+        return ref($id);
+    }
+
+    /**
+     * Creates an inline service.
+     *
+     * @param string|null $class
+     *
+     * @return InlineServiceConfigurator
+     */
+    final public function inline($class = null)
+    {
+        return inline($class);
+    }
+
+    /**
+     * Creates a lazy iterator.
+     *
+     * @param ReferenceConfigurator[] $values
+     *
+     * @return IteratorArgument
+     */
+    final protected function setIterator(array $values)
+    {
+        return iter($values);
+    }
+
+    /**
+     * Creates an expression.
+     *
+     * @param string $expression an expression
+     *
+     * @return Expression
+     */
+    final public function expression($expression)
+    {
+        return expr($expression);
+    }
+}
+
+/**
+ * Creates a service reference.
+ *
+ * @param string $id
+ *
+ * @return ReferenceConfigurator
+ */
+function ref($id)
+{
+    return new ReferenceConfigurator($id);
+}
```

Comment:
Do we need to keep the functions? If functions aren't *the* official way of doing things, nobody will use them. So I'd prefer less options things to maintain, right?

Reply:
see https://github.com/symfony/symfony/pull/23834#issuecomment-326789479

Reply:
So I disagree with keeping them, but not very heavily. So it's ok for me :)

Reply:
I would remove the functions as well :)

Reply:
I agree with Ryan and Fabien. It's better to provide just one solution.